src/queue.rs

```diff
@@ -510,88 +513,51 @@ impl<M: GuestAddressSpace> Queue<M> {
-        let addr = self.driver_area.unchecked_add(2);
+        let addr = self.config().driver_area.unchecked_add(2);
         self.mem
             .memory()
             .load(addr, order)
             .map(Wrapping)
             .map_err(Error::GuestMemory)
     }
 
-    /// A consuming iterator over all available descriptor chain heads offered by the driver.
-    pub fn iter(&mut self) -> Result<AvailIter<'_, M>, Error> {
-        self.avail_idx(Ordering::Acquire).map(move |idx| AvailIter {
-            mem: self.mem.memory(),
-            desc_table: self.desc_area,
-            avail_ring: self.driver_area,
-            last_index: idx,
-            queue_size: self.actual_size(),
-            next_avail: &mut self.next_avail,
-        })
-    }
-
-    /// Puts an available descriptor head into the used ring for use by the guest.
-    pub fn add_used(&mut self, head_index: u16, len: u32) -> Result<(), Error> {
-        if head_index >= self.actual_size() {
-            error!(
-                "attempted to add out of bounds descriptor to used ring: {}",
-                head_index
-            );
-            return Err(Error::InvalidDescriptorIndex);
-        }
-
-        let mem = self.mem.memory();
-        let next_used_index = u64::from(self.next_used.0 % self.actual_size());
-        let addr = self.device_area.unchecked_add(4 + next_used_index * 8);
-        mem.write_obj(VirtqUsedElem::new(head_index, len), addr)
-            .map_err(Error::GuestMemory)?;
-
-        self.next_used += Wrapping(1);
-
-        mem.store(
-            self.next_used.0,
-            self.device_area.unchecked_add(2),
-            Ordering::Release,
-        )
-        .map_err(Error::GuestMemory)
-    }
-
     // Helper method that writes `val` to the `avail_event` field of the used ring, using
     // the provided ordering.
     fn set_avail_event(&self, val: u16, order: Ordering) -> Result<(), Error> {
         let offset = (4 + self.actual_size() * 8) as u64;
-        let addr = self.device_area.unchecked_add(offset);
+        let addr = self.config().device_area.unchecked_add(offset);
         self.mem
             .memory()
             .store(val, addr, order)
             .map_err(Error::GuestMemory)
     }
 
     // Set the value of the `flags` field of the used ring, applying the specified ordering.
-    fn set_used_flags(&mut self, val: u16, order: Ordering) -> Result<(), Error> {
+    fn set_used_flags(&self, val: u16, order: Ordering) -> Result<(), Error> {
         self.mem
             .memory()
-            .store(val, self.device_area, order)
+            .store(val, self.config().device_area, order)
             .map_err(Error::GuestMemory)
     }
 
     // Write the appropriate values to enable or disable notifications from the driver. Every
     // access in this method uses `Relaxed` ordering because a fence is added by the caller
     // when appropriate.
-    fn set_notification(&mut self, enable: bool) -> Result<(), Error> {
```

Comment:
I think these were `mut` by design. In principle they don't need mutable access to `self` because the fields themselves employ interior mutability; however, we don't want multiple threads to muck with a single queue without external protection. The usage of atomic primitives works fine if there is parallel access from the guest, but not if other threads in the VMM call these methods concurrently.

Reply:
That makes sense (and it became even more apparent after Gerry started the discussion about handling queues via multiple threads). I'll make the necessary adjustments after thinking a bit more about the multi-thread angle as well.